### PR TITLE
Configurable API endpoint

### PR DESCRIPTION
**Pull request by nzagorchev**
What              | Where/Who
------------------|----------------------------------------
JIRA Issue        | [SDK-661](https://leanplum.atlassian.net/browse/SDK-661)
People Involved   | @nzagorchev 

## Background
Change API and Dev server endpoints on the fly, based on API response. Response example in `change_host_response.json`. HTTP status will be 200 OK.
In that case, the SDK must retry the request to the new endpoint.

## Implementation
`LPRequestSender` checks if the response returns a new API config, then changes the hosts and retries the request. Callbacks will be executed by the subsequent request.

`LeanplumSocket` is modified, so a new engine can be created and connected with. Changes were required in Socket_IO and WebSocket classes to deallocate objects correctly.
Disconnecting and Connecting again (all async) proved to be inefficient.

_Note:_ `sendNow` for request with datas does _not_ update the API config. Such a request is _only_ sent when uploading app icons.

As part of this PR, the following changes are also made:
- Rewriting the `LPAPIConfig` and separating from `LPConstants`
- Moving user data to User class
- Renaming servlet to path
- Moving Request UUID to Swift

## Testing steps
Manual and Unit tests.

## Is this change backwards-compatible?
Yes

### CODE REVIEW

File: LeanplumSDK/LeanplumSDK/ClassesSwift/User.swift [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+//  User.swift
+//  LeanplumSDK
+//
+//  Created by Nikola Zagorchev on 25.02.22.
+//
+
+import Foundation
+
+@objc public class User: NSObject {
+
```
Comment (milos1290) [inline LeanplumSDK/LeanplumSDK/ClassesSwift/User.swift:11]: remove empty line

File: LeanplumSDK/LeanplumSDK/ClassesSwift/User.swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                let deviceId = self.deviceId
+            else {
+                return nil
+            }
+            
+            return UserKey(appId: appId, userId: userId, deviceId: deviceId)
+        }
+    }
+    
+    public struct UserKey {
```
Comment (milos1290) [inline LeanplumSDK/LeanplumSDK/ClassesSwift/User.swift:47]: Move to the top of the class

File: LeanplumSDK/LeanplumSDK/ClassesSwift/Utilities/WeakTimer.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+//  LeanplumSDK
+//
+//  Created by Nikola Zagorchev on 23.02.22.
+//
+
+import Foundation
+
+@objc(LPWeakTimer)
+public class WeakTimer: NSObject {
+
```
Comment (milos1290) [inline LeanplumSDK/LeanplumSDK/ClassesSwift/Utilities/WeakTimer.swift:12]: remove empty line

File: LeanplumSDK/LeanplumSDK/ClassesSwift/Utilities/WeakTimer.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    
+    @objc(scheduledTimerWithTimeInterval:target:userInfo:repeats:block:)
+    public static func scheduledTimer(timeInterval: TimeInterval,
+                                           target: AnyObject,
+                                           userInfo:Any?,
+                                           repeats: Bool,
+                                           action: @escaping (Timer) -> Void) -> Timer {
+        
+        let target = WeakTimer(target: target, action: action)
+        return Timer.scheduledTimer(timeInterval: timeInterval,
```
Comment (milos1290) [inline LeanplumSDK/LeanplumSDK/ClassesSwift/Utilities/WeakTimer.swift:38]:
```
  return .scheduledTimer(timeInterval: timeInterval,
                                    target: target,
                                    selector: #selector(fire),
                                    userInfo: userInfo,
                                    repeats: repeats)
```

**Review by milos1290 — CHANGES_REQUESTED** [review]

File: LeanplumSDK/LeanplumSDK/ClassesSwift/ApiConfig.swift [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    
+    @objc public static let shared: ApiConfig = .init()
+    
+    @objc public private(set) var appId: String?
+    @objc public private(set) var accessKey: String?
+    
+    @objc public var socketPort = Constants.socketPort
+    @objc public var apiSSL = Constants.apiSSL
+    
+    enum Constants {
```
[the remark this answers is not in the document]
Comment (nzagorchev) [reply LeanplumSDK/LeanplumSDK/ClassesSwift/ApiConfig.swift:21]: Moving to top for now, we can restructure the constants in Dejans PR later on

File: LeanplumSDK/LeanplumSDK/Classes/Managers/Networking/LPRequestSender.m [hunk 6]
```diff
@@ -344,7 +364,32 @@ - (void)sendRequests
     [[LPOperationQueue serialQueue] addOperation:requestOperation];
 }
 
--(void)uiDidTimeout {
+- (BOOL)updateApiConfig:(id)json {
+    if ([json isKindOfClass:NSDictionary.class]) {
```
[the remark this answers is not in the document]
Comment (nzagorchev) [reply LeanplumSDK/LeanplumSDK/Classes/Managers/Networking/LPRequestSender.m:368]: yes, the response should be 200 but with success:false and contain the config. 
The response of the multi is always an array. In this case, it should return only one object with new endpoints but we iterate to be safe.

**Review by hborisoff — APPROVED** [review]
Looks good

**Review by milos1290 — APPROVED** [review]

**Pull request merged**